<?xml version="1.0"?>
<!--
/****************************************************************************
 *   Copyright (c) 2018 Michael Shomin. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 3. Neither the name ATLFlight nor the names of its contributors may be
 *    used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS LICENSE.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * In addition Supplemental Terms apply.  See the SUPPLEMENTAL file.
 ****************************************************************************/
-->
<launch>
  <arg name="8074" default="true"/>

  <group ns="downward">
    <node pkg="nodelet" type="nodelet" name="vio_nodelet_man" args="manager" output="screen"/>

    <!-- camera nodelet -->
    <node pkg="nodelet" type="nodelet" name="snap_cam_nodelet"
       args="load snap_cam_ros/driver vio_nodelet_man">
      <param name="camera_name" value="downward"/>
      <param name="format" value="1"/>
      <param name="camera_number" value="1"/>
      <param name="width" value="640"/>
      <param name="height" value="480"/>
      <param name="frame_id" value="dfc"/>
      <param name="is_cam_master" value="true"/>
    </node>

    <param name="image_raw/compressed/jpeg_quality" value="50"/>

    <!-- VIO nodelet -->
    <node pkg="nodelet" type="nodelet" name="snap_vio_nodelet"
          args="load snap_vio/driver vio_nodelet_man" output="screen">
      <remap from="camera_info" to="fisheye/camera_info"/>

      <param name="delta" value="-0.0068"  if="$(arg 8074)"/>
      <param name="delta" value=".002" unless="$(arg 8074)"/>
      <param name="snav_mode" value="true"/>
    </node>

    <!-- fisheye camera info spoofer -->
    <include file="$(find snap_vio)/launch/fisheye_camera_info.launch"/>

    <!-- imu nodelet -->
    <node pkg="nodelet" type="nodelet" name="snap_imu_nodelet"
          args="load snap_imu/driver vio_nodelet_man" output="screen">
    </node>

    <!-- imu nodelet -->
    <node pkg="nodelet" type="nodelet" name="snap_cpa_nodelet"
          args="load snap_cpa/driver vio_nodelet_man" output="screen">
    </node>

    <!-- SNAV VIO injector node -->
    <node pkg="snap_vio" type="snav_vio_injector_node"
	  name="snav_vio_injector" output="screen"/>

    <!-- SNAV ROS -->
    <node pkg="snav_ros" type="snav_interface_node"
	  name="snav_interface_node" output="screen"/>
  </group>

  <!-- Board/Vehicle URDF -->
  <include file="$(find qflight_descriptions)/launch/qflight_robot.launch">
    <arg name="board" value="sdf_tray_45"/>
    <arg name="vehicle" value="230"/>
    <arg name="imu_is_root_tf" value="false"/>
  </include>

</launch>
